### ISSUE TITLE
Environments always recreated, except for version matching env interpreter

### ISSUE THREAD

**Issue by JarnoRFB** [tox-dev/tox-conda#20]
Hi,
I just tried it out on Ubuntu 18.10. I have a conda env with py36. Now I want to test against 3.5 and 3.7 as well. My `tox.ini` looks like this.

```
[tox]
envlist = py35, py36, py37, flake8

[testenv]
deps =
    -rrequirements-dev.txt
commands =
    pytest \
        {posargs}
```

It works for all versions, which is great. However, the environment seems only to be cached for for 3.6 and not for 3.5 and 3.7. While the environments stay on the file system, they are not reused. 

```
$ ls .tox
dist/  log/  py35/  py36/  py37/
```

When running `tox`, I get output like `py35 recreate:`  . What could be the reason for that?

**Comment by rsokl** [comment]
I have a suspicion that a [bug I encountered](https://github.com/tox-dev/tox/issues/1259) has the same root-cause as your issue. 

If you run `tox -v`, do you see something like:

```
write config to /home/yourname/test/.tox/py35/.tox-config1 as '00000000000000000000000000000000 None\n3.8.6 0 0 0'
write config to /home/yourname/test/.tox/py36/.tox-config1 as 'c96b4313481accc74162889f429b1f0d /home/yourname/anaconda3/bin/python\n3.8.6 0 0 0'
write config to /home/yourname/test/.tox/py37/.tox-config1 as '00000000000000000000000000000000 None\n3.8.6 0 0 0'
```

It looks like the plugin is unable to find the path to the source executable for some of its environments at an early stage of the tox process. I suspect that this prevents it from reusing the cached environments. However it somehow recovers gracefully enough to actually build these environments.

**Comment by gaborbernat** [comment]
Indeed it seems conda did not overwrite the interpreter discovery mechanism, we'll need to fix this.

**Comment by JarnoRFB** [comment]
@rsokl Running `tox -v` starts to recreate the environments and gives output like

```
using tox.ini: /home/jarno/projects/incense/tox.ini (pid 14855)
using tox-3.8.4 from /home/jarno/.miniconda/envs/incense-dev/lib/python3.6/site-packages/tox/__init__.py (pid 14855)
GLOB sdist-make: /home/jarno/projects/incense/setup.py
[14859] /home/jarno/projects/incense$ /home/jarno/.miniconda/envs/incense-dev/bin/python setup.py sdist --formats=zip --dist-dir .tox/dist >.tox/log/GLOB-0.log
package .tmp/package/1/incense-0.0.7+11.g2581127.zip links to dist/incense-0.0.7+11.g2581127.zip (/home/jarno/projects/incense/.tox)
py35 cannot reuse: attr base_resolved_python_path 'None'!=None
py35 recreate: /home/jarno/projects/incense/.tox/py35
[14877] /home/jarno/projects/incense/.tox$ /home/jarno/.miniconda/bin/conda create --yes -p py35 python=3.5 >py35/log/py35-0.log
py35 installdeps: -rrequirements-dev.txt
```

but great if you already spotted the source of the problem!

**Comment by andrew-schweitzer-analog** [comment]
I _think_ I was able to prevent unnecessary environment recreation by adding these settings: 
```
[tox]
ignore_basepython_conflict = True

[testenv]
basepython = python
```

I am running tox/tox-conda from a 3.7 conda environment. I run mypy, pylint, py37-conda, py37-pypi, py38-conda and py38-pypi environments in tox. Without setting `ignore_basepython_conflict` and `basepython` as above, mypy, pylint, py37-conda and py37-pypi _reused_ the environment, but py38-conda and py38-pypi _recreated_ the environment. With the above settings all environments are reused.

I _think_ what is going on is that without tox-conda, tox tries to 'find' an appropriate system python to build each environment, but that _with_ tox-conda environments are not created _with_ python (directly) but with conda instead. With tox-conda we don't _want_ to find a system python; we just want to tell conda what python version to use when creating the environment. 

@gaborbernat, do you have an idea of tox-conda should overwrite in the interpreter discovery mechanism? Maybe tox-conda should just document that you should essentially disable basepython, as above?

**Comment by andrew-schweitzer-analog** [comment]
This is probably obvious to the maintainers, but relevant sections of code seem to include: 
* [config.basepython_default()](https://github.com/tox-dev/tox/blob/master/src/tox/config/__init__.py#L576)
* [unix.tox_get_python_executable()](https://github.com/tox-dev/tox/blob/master/src/tox/interpreters/unix.py#L10)
* [tox-conda.tox_get_python_executable()](https://github.com/tox-dev/tox-conda/blob/master/tox_conda/plugin.py#L161)
* [common.base_discover()](https://github.com/tox-dev/tox/blob/master/src/tox/interpreters/common.py#L7)
* [py_path.check_with_path()](https://github.com/tox-dev/tox/blob/121ef6ba07c0a8e6302e50aabca16c747f036fff/src/tox/interpreters/via_path.py#L17)
* [Interpreters.get_executable](https://github.com/tox-dev/tox/blob/master/src/tox/interpreters/__init__.py#L18)
* [decision to re-use](https://github.com/tox-dev/tox/blob/master/src/tox/venv.py#L245)
* [write-back of python used](https://github.com/tox-dev/tox/blob/master/src/tox/venv.py#L304)

**Comment by grubbins** [comment]
I ran into this issue today. @andrew-schweitzer-analog , your workaround made no difference for me: the base python version is still recorded as `None` in all the created envs where the python version doesn't match the python version of the environment from which I ran tox...

**Comment by andrew-schweitzer-analog** [comment]
It also broke again for me, this time testing python 3.6, rather than mypy/pylint. I've been tasks like mypy and pylint that are not explicitly testing the code against different python versions/environments out of tox and into make to save regeneration time.

**Comment by AntoineD** [comment]
There is a fix in the master branch for this, could you give it a try?

**Comment by rwhitt2049** [comment]
@AntoineD, I'm still having this issue on Ubuntu (although 0.4.0 fixed Windows). I'm running WSL Ubuntu 20.04, and tox-conda=0.4.0

**Comment by AntoineD** [comment]
It should be fixed in #53.

**Comment by gaborbernat** [comment]
Release on-going as 0.4.1 under https://github.com/tox-dev/tox-conda/runs/1380704093
